Add implementation and test for calculate_nft_ownership_layer_puzzle_hash

greimela opened the pull request:
Plus some review comments

greimela committed:
Add implementation and test for calculate_nft_ownership_layer_puzzle_hash

greimela committed:
Add some review comments

trgarrett commented on clsp/fusion_singleton.clsp:
```diff
@@ -39,26 +39,22 @@
       (list ASSERT_MY_COIN_ID singleton_coin_id)
       (list ASSERT_MY_PUZZLEHASH (calculate_full_puzzle_hash SINGLETON_MOD_HASH LAUNCHER_ID LAUNCHER_PUZZLE_HASH singleton_inner_puzzlehash))
-      (list ASSERT_MY_AMOUNT 1)
+      (list ASSERT_MY_AMOUNT 1) ; This feels unnecessary. And if you ever create such a singleton with amount > 1 it is locked forever.
```
Thanks.

trgarrett commented on clsp/fusion_singleton.clsp:
```diff
@@ -39,26 +39,22 @@
-          )
-        )
-        (c 
+  ; You can simplify those for loops a lot by turning them upside down:
```
I feel dense, but I'm not quite following here. Could you describe the mechanism a bit more?

trgarrett commented on clsp/utils.clib:
```diff
@@ -60,14 +60,14 @@
+  (defun-inline calculate_nft_ownership_layer_puzzle_hash (NFT_OWNERSHIP_LAYER_MOD_HASH current_owner transfer_program_hash inner_puzzle_hash)
+      (curry_hashes NFT_OWNERSHIP_LAYER_MOD_HASH
+        (sha256 1 NFT_OWNERSHIP_LAYER_MOD_HASH)
+        (sha256 1 current_owner)
```
Thanks. I like the current_owner terminology better. I was trying to base it on what the ownership layer called it, but I agree it didn't make sense.

greimela replied:
When I look at `nft_ownership_layer.clsp` in the chia-blockchain repo, they call that parameter CURRENT_OWNER too

trgarrett replied:
Which branch are you on? I see new_owner a bunch of places.

trgarrett commented on clsp/p2_fusion.clsp:
```diff
@@ -22,7 +22,7 @@
         (list
             ; prove the parentage of the supplied NFT coin matches the expected derived puzzle hash
             (list ASSERT_MY_PUZZLEHASH (calculate_full_puzzle_hash SINGLETON_MOD_HASH nft_launcher_id LAUNCHER_PUZZLE_HASH nft_inner_puzzle_hash))
-            (list ASSERT_MY_AMOUNT 1)
```
If I prove the NFT is the one I desire by virtue of its full puzzlehash, is there any value in asserting amount for the NFT at all? I am hard-coding it some places in my Python driver to 1, but if we encounter non-1 values in the wild (which I do realize are possible), I could update the Python driver without any blockchain impacts.

trgarrett replied:
Ewww...another potential corner case maybe?

Would it be possible to send standard XCH to the extended puzzlehash and convince the p2 singleton that it was an NFT? Thoughts @greimela @yakuhito @BrandtH22

Yakuhito replied:
The launcher ids are curried in - so you must have a singleton with the NFT launcher id. This is safe IMO.

greimela committed:
Very subjective refactoring of the fusion_singleton.clsp

- use SINGLETON_STRUCT
- use a list of tuples for the nfts to lock/unlock
- Only two loops, one for the locked NFTs and one for the unlocked NFTs

Yakuhito commented on clsp/fusion_singleton.clsp:
```diff
@@ -31,149 +26,98 @@
 
       ; spend myself forward
-      (list CREATE_COIN singleton_inner_puzzlehash 1 (list LAUNCHER_ID)) ;; opt to keep details beyond launcher ID encoded in eve coin spend
+      (list CREATE_COIN singleton_inner_puzzlehash 1 (list (f (r SINGLETON_STRUCT)))) ;; opt to keep details beyond launcher ID encoded in eve coin spend
```
is this needed? can always be found in the parent's puzzle reveal

greimela replied:
The puzzle doesn't need it. @trgarrett, are you planning to use the hint for easier lookups?

trgarrett replied:
Yes, I actually intend to update the hint with some information that helps
1.) find fusion puzzles, generally 2.) find what project is claiming them.
People could be a nuisance and put garbage hints on other coins, but the
set of coins that are correct will be at most the size of all coins with
those hints.

trgarrett commented:
Yes, I actually intend to update the hint with some information that helps
1.) find fusion puzzles, generally 2.) find what project is claiming them.
People could be a nuisance and put garbage hints on other coins, but the
set of coins that are correct will be at most the size of all coins with
those hints.

On Fri, Dec 8, 2023 at 5:57 AM Andreas Greimel ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In clsp/fusion_singleton.clsp
> <https://github.com/trgarrett/fusion-clsp/pull/4#discussion_r1420339207>:
>
> >
>        ; spend myself forward
> -      (list CREATE_COIN singleton_inner_puzzlehash 1 (list LAUNCHER_ID)) ;; opt to keep details beyond launcher ID encoded in eve coin spend
> +      (list CREATE_COIN singleton_inner_puzzlehash 1 (list (f (r SINGLETON_STRUCT)))) ;; opt to keep details beyond launcher ID encoded in eve coin spend
>
> The puzzle doesn't need it. @trgarrett <https://github.com/trgarrett>,
> are you planning to use the hint for easier lookups?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/trgarrett/fusion-clsp/pull/4#discussion_r1420339207>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI2B4MUMNERD4Z7BEEDFTJ3YIL6D5AVCNFSM6AAAAABAK2T2AKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTONZSGE4TQMBVGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

trgarrett commented on clsp/fusion_singleton.clsp:
```diff
@@ -31,149 +26,98 @@
+          (assert_p2_condition SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH (f nfts_to_unlock) OFFER_MOD_HASH)
+          (c
+            (announce_p2_condition SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH (f nfts_to_unlock) singleton_coin_id OFFER_MOD_HASH)
+            (compute_nft_unlock_conditions SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH OFFER_MOD_HASH (r nfts_to_unlock) singleton_coin_id offer_nonce)
```
I'm a little baffled as to why this function takes a conditions_so_far but doesn't handle it in the recursive call. Can you help explain @greimela ?

greimela replied:
~~I'm confused, my local code looks different and I don't know why the remote code doesn't mirror it.~~

Ah, you fixed it already 😅 
Your changes look good, I'm sorry for the state my code was in...
